### PR TITLE
Extend type code generator

### PR DESCRIPTION
**Pull request by turbolent**
Depends on #2805 

## Description

- Add support for constructors/initializers, also in the nested case (requires defining member in container type)
- Refactor tests so they are generated into separated packages, and import them, so they are actually type checked, and can be tested with the existing parsing/checking testing infrastructure
- Do not automatically add entitlements to the base type activation. This isn't done for any other type declaration, and pollutes the base type activation when the types are not generated into sema.
  We might want to solve this in a general case for all type declaration in a follow-up PR.
- Properly refer to built-in types in the `sema` package. For example, if `Int` is mentioned in the `cdc` file, generate `sema.IntType` when the type code output is not generated into `sema`.

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/sema/gen/golden_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          let x = Foo(bar: 1)
+        `,
+		checker.ParseAndCheckOptions{
+			Config: &sema.Config{
+				BaseValueActivation: baseValueActivation,
+			},
+		},
+	)
+	require.NoError(t, err)
+}
```
Comment (turbolent) [inline runtime/sema/gen/golden_test.go:71]: The new package structure allows us to test the generated type declarations with the type checker

File: runtime/sema/gen/golden_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/equatable"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/exportable"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/fields"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/functions"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/importable"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/member_accessible"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/nested"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/simple_resource"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/simple_struct"
+	_ "github.com/onflow/cadence/runtime/sema/gen/testdata/storable"
```
Comment (turbolent) [inline runtime/sema/gen/golden_test.go:44]: Import the "generated" code from all test cases, so it is type checked.

This ensures the type code generator does not generate code that does not actually work.

File: runtime/sema/gen/main.go [hunk 3]
```diff
@@ -323,8 +326,106 @@ func (g *generator) declarationDocString(decl ast.Declaration) dst.Expr {
 	return renderDocString(docString)
 }
 
-func (*generator) VisitSpecialFunctionDeclaration(_ *ast.SpecialFunctionDeclaration) struct{} {
-	panic("special function declarations are not supported")
+func (g *generator) VisitSpecialFunctionDeclaration(decl *ast.SpecialFunctionDeclaration) (_ struct{}) {
```
Comment (turbolent) [inline runtime/sema/gen/main.go:329]: Main contribution: add support for initializers

File: runtime/sema/gen/main.go [hunk 4]
```diff
@@ -527,65 +630,93 @@ func (g *generator) VisitCompositeDeclaration(decl *ast.CompositeDeclaration) (_
...
-			members := membersExpr(typeDecl.fullTypeName, tyVarName, memberDeclarations)
+			members := membersExpr(
+				fullTypeName,
+				tyVarName,
+				memberDeclarations,
+			)
 
 			const membersVariableIdentifier = "members"
 
+			stmts := []dst.Stmt{
```
Comment (turbolent) [inline runtime/sema/gen/main.go:644]: Moved up from below

File: runtime/sema/gen/main.go [hunk 5]
```diff
@@ -1020,7 +1171,14 @@ func functionTypeExpr(
 	if t.ReturnTypeAnnotation != nil {
 		returnTypeExpr = typeExpr(t.ReturnTypeAnnotation.Type, typeParams)
 	} else {
-		returnTypeExpr = typeVarIdent("Void")
```
Comment (turbolent) [inline runtime/sema/gen/main.go:1023]: Needed, to properly package-qualified `sema.VoidType`, if needed

File: runtime/stdlib/builtin.go [hunk 6]
```diff
@@ -18,10 +18,6 @@
 
 package stdlib
 
-import "github.com/onflow/cadence/runtime/sema"
-
-var UInt8Type = sema.UInt8Type
```
Comment (turbolent) [inline runtime/stdlib/builtin.go:23]: Such declarations are no longer needed

**Review by dsainati1 — APPROVED** [review]

**Review by SupunS — APPROVED** [review]

File: runtime/sema/account.gen.go [hunk 7]
```diff
@@ -2099,57 +2099,30 @@ var CapabilitiesMappingType = &EntitlementMapType{
 
 func init() {
 	BuiltinEntitlementMappings[AccountMappingType.Identifier] = AccountMappingType
-	addToBaseActivation(AccountMappingType)
```
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/sema/account.gen.go:2102]: No, this isn't related to adding support for initializers/constructors.

The generator produced this and the statement above literally: always register the entitlement (mapping) type in `BuiltinEntitlements` / `BuiltinEntitlementMappings`, and register the type in the base activation. 

However, there are a couple problems with this: This approach assumes the code is generated into the `sema` package. First, this is not always the case: The code generator may be used outside of `sema`, e.g. to generate type code for the standard library (`stdlib` package). There, `addToBaseActivation` is neither available (it is a private/unexported function), nor should the type be defined in the **global** base activation.

This became apparent when I refactored the test cases to not generated into a package named `stdlib` and made the test code importable, but instead generated it into a package named with the test name (see packages in `testdata`)

[the remark this answers is not in the document]
Comment (SupunS) [reply runtime/sema/account.gen.go:2102]: ah ok, thanks for the details 👍

**Pull request merged**